boa/src/syntax/lexer/operator.rs

```diff
@@ -0,0 +1,159 @@
+use std::io::Read;
+
+/// `vop` tests the next token to see if we're on an assign operation of just a plain binary operation.
+///
+/// If the next value is not an assignment operation it will pattern match  the provided values and return the corresponding token.
+macro_rules! vop {
+    ($cursor:ident, $assign_op:expr, $op:expr) => ({
+        match $cursor.peek()? {
+            None => Err(Error::syntax("Abrupt end: could not preview next value as part of operator")),
+            Some('=') => {
+                $cursor.next()?.expect("= token vanished");
+                $cursor.next_column();
+                $assign_op
+            }
+            Some(_) => $op,
+        }
+    });
+    ($cursor:ident, $assign_op:expr, $op:expr, {$($case:pat => $block:expr), +}) => ({
+        // let punc = $cursor.peek().ok_or_else(|| Error::syntax("could not preview next value"))?;
+        match $cursor.peek()? {
+            None => Err(Error::syntax("Abrupt end: could not preview next value as part of operator")),
+            Some('=') => {
+                $cursor.next()?.expect("= token vanished");
+                $cursor.next_column();
+                $assign_op
+            },
+            $($case => {
+                $cursor.next()?.expect("Token vanished");
+                $cursor.next_column();
+                $block
+            })+,
+            _ => $op,
+        }
+    });
+    ($cursor:ident, $op:expr, {$($case:pat => $block:expr),+}) => {
+        match $cursor.peek().ok_or_else(|| LexerError::syntax("could not preview next value"))? {
+            $($case => {
+                $cursor.next()?;
+                $cursor.next_column();
+                $block
+            })+,
+            _ => $op
+        }
+    }
+}
+
+/// The `op` macro handles binary operations or assignment operations and converts them into tokens.
+macro_rules! op {
+    ($cursor:ident, $start_pos:expr, $assign_op:expr, $op:expr) => ({
+        Ok(Token::new(
+            vop!($cursor, $assign_op, $op)?.into(),
+            Span::new($start_pos, $cursor.pos()),
+        ))
+    });
+    ($cursor:ident, $start_pos:expr, $assign_op:expr, $op:expr, {$($case:pat => $block:expr),+}) => ({
+        let punc: Punctuator = vop!($cursor, $assign_op, $op, {$($case => $block),+})?;
+        Ok(Token::new(
+            punc.into(),
+            Span::new($start_pos, $cursor.pos()),
+        ))
+    });
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(super) struct Operator {
+    init: char,
+}
+
+/// Operator lexing.
+///
+/// Assumes that the cursor has already consumed the operator starting symbol (stored in init).
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///  - [MDN documentation][mdn]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-multiplicative-operators &
+///         https://tc39.es/ecma262/#sec-additive-operators &
+///         https://tc39.es/ecma262/#sec-relational-operators &
+///         https://tc39.es/ecma262/#sec-bitwise-shift-operators
```

Comment:
I don't think this is valid markdown we should use one link

Reply:
Fixed